File: templates/java/src/test/java/com/mbreissi/testcomponent/AppTest.java
package <<PACKAGE>>;

import com.mbreissi.edgecommons.commands.CommandException;
import com.mbreissi.edgecommons.messaging.Message;
import com.mbreissi.edgecommons.messaging.MessageBuilder;
import com.google.gson.JsonObject;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * The connectivity this component reports, and the {@code set-greeting} command handler.
 *
 * <p>A service owns no southbound connections, so it reports <b>no instances</b> — and that is the
 * contract, not an omission: the {@code state} keepalive carries no {@code instances[]} section, and
 * the built-in {@code status} verb therefore answers exactly as {@code ping} does
 * ({@code {"status":"RUNNING","uptimeSecs":n}}). Pinned here so it stays deliberate: the day this
 * component acquires a connection, this test is what tells you to report it.
 *
 * <p>The one piece of real logic — the {@code set-greeting} verb's parse/validate/swap — lives in
 * {@link Greeting}, out of the live run loop, so it is unit-tested here without a broker.
 */
class <<COMPONENTNAME>>Test {

    private static Message body(JsonObject payload) {
        return MessageBuilder.create("set-greeting", "1.0").withPayload(payload).build();
    }

    private static JsonObject greeting(String value) {
        JsonObject o = new JsonObject();
        o.addProperty("greeting", value);
        return o;
    }

    @Test
    void aComponentWithNoConnectionsReportsNoInstances() {
        assertTrue(<<COMPONENTNAME>>.instanceConnectivity().isEmpty());
    }

    @Test
    void theGreetingStartsAtItsInitialValue() {
        assertEquals("Hi", new Greeting("Hi").get());
    }

    @Test
    void setGreetingSwapsTheValueAndReportsThePrevious() throws CommandException {
        Greeting g = new Greeting("first");

        JsonObject reply = g.apply(body(greeting("second")));

        assertEquals("first", reply.get("previousGreeting").getAsString(),
                "the reply echoes what it replaced, so a console sees the effect");
        assertEquals("second", reply.get("greeting").getAsString());
        assertEquals("second", g.get(), "the next status tick reflects the new greeting");
    }

    @Test
    void aMissingOrMalformedGreetingIsBadArgs() {
        Greeting g = new Greeting("start");

        // No `greeting` field at all.
        assertEquals("BAD_ARGS", codeOf(g, body(new JsonObject())));
        // Present but not a primitive (an object, not a string/number).
        JsonObject nested = new JsonObject();
        nested.add("greeting", new JsonObject());
        assertEquals("BAD_ARGS", codeOf(g, body(nested)));
        // A non-JSON body is rejected the same way — the state is untouched.
        assertEquals("BAD_ARGS", codeOf(g, MessageBuilder.create("set-greeting", "1.0")
                .withPayload("not-json").build()));
        assertEquals("start", g.get(), "a rejected command must not mutate the greeting");
    }

    private static String codeOf(Greeting g, Message request) {
        try {
            g.apply(request);
        } catch (CommandException e) {
            return e.getCode();
        }
        throw new AssertionError("expected the command to fail with BAD_ARGS");
    }
}

File: templates/typescript/test/app.test.ts
import { CommandException } from "@edgecommons/edgecommons";
import { describe, expect, it } from "vitest";

import { INITIAL_GREETING, applyGreeting, instanceConnectivity } from "../src/app";

describe("the instance-connectivity provider", () => {
  it("reports no instances, because this component owns no connections", () => {
    // The provider the `state` keepalive pushes and the `status` verb pulls — one source, two
    // surfaces. Reporting nothing is the contract, not an omission: with no instances the keepalive
    // carries no `instances[]` section and `status` answers exactly as `ping` does.
    expect(instanceConnectivity()).toEqual([]);
  });
});

describe("the set-greeting command decision", () => {
  it("computes the greeting change from a well-formed body", () => {
    // The runtime applies `.greeting` to its in-memory state and returns the whole object, so a
    // console sees the previous and new greeting and the next status tick reflects the new one.
    const change = applyGreeting({ greeting: "hi there" }, INITIAL_GREETING);
    expect(change).toEqual({ previousGreeting: INITIAL_GREETING, greeting: "hi there" });
  });

  it("rejects a body without a string `greeting` with BAD_ARGS", () => {
    // A malformed command must be refused, not silently ignored — the caller gets a typed reason.
    for (const bad of [undefined, null, 42, "nope", {}, { greeting: 7 }]) {
      expect(() => applyGreeting(bad, INITIAL_GREETING)).toThrow(CommandException);
    }
  });
});
